### PR TITLE
ONNX export: Fully connected operator w/o bias, ReduceSum, Square

### PR DESCRIPTION
**Pull request by vandanavk**
## Description ##
Fully connected operator can be used with no bias. When bias is not specified, making use of bias 0. 
Enabling export of Square operator and Sum operator

v2: Based on @zhreshold's inputs, added a helper function to convert bool (string) attrs to int
v3: Addressed @Roshrini's and @anirudhacharya's review comments
v4: Added Roshani's commit for importing square operator. Added a test for square operator.
v5: Removed import square based on an offline discussion. Instead, based on Roshani's suggestion, modified existing import of Power operator to be compatible with all versions of ONNX's Pow operator.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- create a bias tensor with default value 0 when no_bias=True
- Export mx.sym.Square using Power operator in ONNX
- Export mx.sym.Sum using ReduceSum operator in ONNX
- Added tests for ReduceSum

## Comments ##
- tested with a test model which has both types of nodes - fully connected with bias and fully connected without bias.
- Tested onnx_backend_test.py

### CODE REVIEW

**Comment by vandanavk** [comment]
@Roshrini @saswatac

**Comment by vandanavk** [comment]
@mxnet-label-bot [pr-awaiting-review]

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 1]
```diff
@@ -212,20 +212,46 @@ def convert_fully_connected(node, **kwargs):
...
-    bias_node = proc_nodes[bias_node_id]
 
+    input_node = proc_nodes[input_node_id]
     input_name = input_node.name
+
+    weights_node = proc_nodes[weight_node_id]
     weights_name = weights_node.name
-    bias_name = bias_node.name
+
+    if no_bias == "False":
```
Comment (Roshrini) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: Can you check if no_bias is encoded as string '0' or 'False', if both are allowed then need to make changes.

**Comment by anirudhacharya** [comment]
can you please update the PR title to reflect that this PR is only changing the ONNX translation of FullyConnected operator and not the FullyConnected operator itself?

**Comment by vandanavk** [comment]
@Roshrini @anirudhacharya @zhreshold Rebased on top of PR https://github.com/apache/incubator-mxnet/pull/12633

**Comment by Roshrini** [comment]
Looks good to me. @zhreshold Can you please take a look once?

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 2]
```diff
@@ -214,17 +214,42 @@ def convert_fully_connected(node, **kwargs):
     onnx = import_onnx_modules()
     name = node["name"]
     inputs = node["inputs"]
+    attrs = node["attrs"]
+    initializer = kwargs["initializer"]
+
+    no_bias = 1 if ("no_bias" in attrs) and \
```
Comment (zhreshold) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: please functionize all similar calls to improve readability and reduce redundancy

Comment (vandanavk) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: there are 3 such checks (including this) in the current code. Will create a common function. Thanks.

**Comment by vandanavk** [comment]
@zhreshold @Roshrini @anirudhacharya review comments addressed. please have a look.

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 3]
```diff
@@ -127,6 +127,14 @@ def convert_string_to_list(string_val):
 
     return result_list
 
+def convert_bool_to_int(attrs, attr_name):
```
Comment (anirudhacharya) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: you are not converting boolean values to int. you converting string values to int. I think this should be named differently, not sure what though

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 4]
```diff
@@ -753,7 +789,7 @@ def convert_pooling(node, **kwargs):
     pool_types = {"max": "MaxPool", "avg": "AveragePool"}
     global_pool_types = {"max": "GlobalMaxPool", "avg": "GlobalAveragePool"}
 
-    if global_pool:
+    if global_pool == 1:
```
Comment (anirudhacharya) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: `if global_pool` should still be fine. why this change?

**Comment by vandanavk** [comment]
Added the commit from PR https://github.com/apache/incubator-mxnet/pull/12653 here.

**Comment by sandeep-krishnamurthy** [comment]
@zhreshold @anirudhacharya @Roshrini - Can you please take a look. Your feedbacks are addressed. Thanks.

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py (hunk 3 above)
Comment (anirudhacharya) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: you can call it `get_attribute_value`

Comment (vandanavk) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: `get_boolean_attribute_value`? because this helper gets only bool valued attributes

Comment (anirudhacharya) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: ok

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 5]
```diff
@@ -2077,3 +2112,79 @@ def convert_sqrt(node, **kwargs):
...
+    onnx = import_onnx_modules()
+    name = node["name"]
+    proc_nodes = kwargs["proc_nodes"]
+    inputs = node["inputs"]
+
+    input_node_a_id = kwargs["index_lookup"][inputs[0][0]]
+    input_node_a = proc_nodes[input_node_a_id].name
+
+    initializer = kwargs["initializer"]
+    np_arr = np.array([2])
```
Comment (anirudhacharya) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: can you please remove this line

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 6]
```diff
@@ -2077,3 +2112,79 @@ def convert_sqrt(node, **kwargs):
...
+    data_type = onnx.mapping.NP_TYPE_TO_TENSOR_TYPE[np_arr.dtype]
```
Comment (anirudhacharya) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: use np.int64 here

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 7]
```diff
@@ -2077,3 +2112,79 @@ def convert_sqrt(node, **kwargs):
...
+    dims = np.shape(np_arr)
```
Comment (anirudhacharya) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: can this be removed?

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 8]
```diff
@@ -2077,3 +2112,79 @@ def convert_sqrt(node, **kwargs):
...
+
+    power2_name = "square_tensor" + str(kwargs["idx"])
+    tensor_node = onnx.helper.make_tensor_value_info(power2_name, data_type, dims)
+    initializer.append(
+        onnx.helper.make_tensor(
+            name=power2_name,
+            data_type=data_type,
+            dims=dims,
```
Comment (anirudhacharya) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: use tuple `(1,)` here

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 9]
```diff
@@ -2077,3 +2112,79 @@ def convert_sqrt(node, **kwargs):
...
+
+    mx_axis = attrs.get("axis", None)
+    axes = convert_string_to_list(str(mx_axis)) if mx_axis is not None else None
+
+    keepdims = get_boolean_attribute_value(attrs, "keepdims")
+
+    input_node_id = kwargs["index_lookup"][inputs[0][0]]
+    input_node = proc_nodes[input_node_id].name
+
+    if axes is not None:
```
Comment (anirudhacharya) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: can this be `if axes`

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 10]
```diff
@@ -214,17 +222,44 @@ def convert_fully_connected(node, **kwargs):
...
+
+    fcnode = []
+
+    if no_bias == 0:
+        bias_node_id = kwargs["index_lookup"][inputs[2][0]]
+        bias_node = proc_nodes[bias_node_id]
+        bias_name = bias_node.name
+    else:
+        np_arr = np.array([0])
+        data_type = onnx.mapping.NP_TYPE_TO_TENSOR_TYPE[np_arr.dtype]
```
Comment (anirudhacharya) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: please use `np.int64` here and remove the previous line

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 11]
```diff
@@ -214,17 +222,44 @@ def convert_fully_connected(node, **kwargs):
...
+        dims = np.shape(np_arr)
+        bias_name = "bias" + str(kwargs["idx"])
+        tensor_node = onnx.helper.make_tensor_value_info(bias_name, data_type, dims)
+        initializer.append(
+            onnx.helper.make_tensor(
+                name=bias_name,
+                data_type=data_type,
+                dims=dims,
```
Comment (anirudhacharya) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: you can directly use the tuple value `(1,)` and remove the dims variable.

**Comment by vandanavk** [comment]
@anirudhacharya Addressed the comments above. Working on the test for square export.

**Comment by vandanavk** [comment]
> @anirudhacharya Addressed the comments above. Working on the test for square export.

Added a test for square operator - since the test includes both export and import, added the test only in mxnet_export_test.py

**Comment by vandanavk** [comment]
Addressed all the review comments so far. Please have a look @zhreshold @anirudhacharya @Roshrini
